aggregator/src/cache.rs

```diff
@@ -206,3 +210,264 @@ impl PeerAggregatorCache {
+            report_writer: Arc::new(report_writer),
+            entries: Default::default(),
+            taskprov_enabled,
+            ttl_seconds: Uniform::new(ttl - jitter, ttl + jitter),
+        }
+    }
+
+    /// Calculates the next possible cache TTL. This contains jitter to prevent all tasks in the
+    /// cache from falling on the same deadline on application startup.
+    fn next_ttl(&self) -> Time {
+        self.clock.now().saturating_add(&Duration::from_seconds(
+            thread_rng().sample(self.ttl_seconds),
+        ))
+    }
+
+    pub async fn get(&self, task_id: &TaskId) -> Result<Option<Arc<TaskAggregator<C>>>, Error> {
+        // Step one: grab the existing entry for this task, if one exists. If there is no existing
+        // entry, write a new (empty) entry.
+        let entry = {
+            // Unwrap safety: mutex poisoning.
+            let mut entries = self.entries.lock().unwrap();
+            Arc::clone(entries.entry(*task_id).or_insert(Arc::new(Mutex::new(
+                TaskAggregatorCacheEntry {
+                    expiration: self.next_ttl(),
+                    is_initialized: false,
+                    task_aggregator: None,
+                },
+            ))))
+        };
+
+        // Step two: if the entry is empty, fill it via a database query. Concurrent callers
+        // requesting the same task will contend over this lock while awaiting the result of the DB
+        // query, ensuring that in the common case only a single query will be made for each task.
+        let task_aggregator = {
+            let mut entry = entry.lock().await;
+            if self.clock.now() > entry.expiration || !entry.is_initialized {
+                *entry = TaskAggregatorCacheEntry {
+                    expiration: self.next_ttl(),
+                    is_initialized: true,
+                    task_aggregator: self
+                        .datastore
+                        .run_tx("task_aggregator_get_task", |tx| {
+                            let task_id = *task_id;
+                            Box::pin(async move { tx.get_aggregator_task(&task_id).await })
+                        })
+                        .await?
+                        .map(|task| TaskAggregator::new(task, Arc::clone(&self.report_writer)))
+                        .transpose()?
+                        .map(Arc::new),
+                };
+            }
+            entry.task_aggregator.clone()
+        };
+
+        // If taskprov is enabled, we can never cache a None result, because any aggregator replica
+        // can at any time insert a new task and expect it to be immediately available across all
+        // replicas. Remove the None result.
+        //
+        // This unfortunately means that for taskprov cases, tasks that are truly non-existent will
+        // always go down the slow path. It would be up to rate-limiting to mitigate this.
+        if self.taskprov_enabled && task_aggregator.is_none() {
```

Comment:
If `taskprov_enabled` is false, we'll never remove anything from the cache, opening the potential for DoS via random-task-ID-spraying.

I suspect this will be fixed by implementing a cache-eviction strategy.